feat(normalisation): jsonNormalisation/v3 and fixes to jsonNormalisation/v1 as well as jsonNormalisation/v2

jakobmoellerdev opened the pull request:
<!-- markdownlint-disable MD041 -->
#### What this PR does / why we need it

Package jsonv3 provides a normalization which is completely based on the abstract (internal) version of the component descriptor and is therefore agnostic of the final serialization format. Signatures using this algorithm can be transferred among different schema versions, as long as is able to handle the complete information using for the normalization. jsonv2 is the predecessor of this version but had internal defaulting logic that is no longer included as part of this normalization. Thus v3 should be preferred over v2. Note that between v2 and v3 differences can occur mainly if the "extra identity" field is not unique, in which case the v2 normalization opinionated on how to differentiate these items. This no longer happens in v3, meaning the component descriptor is normalized as is.

v2 and v1 were adjusted to accomodate the old(but new because forgotten) legacy behavior in legacy.go. Without this, old signatures would not work. This means this should be (at least partially) back-ported to the last minor versioned released without this correction.

#### Which issue(s) this PR fixes
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
This issue fixes https://github.com/open-component-model/ocm/issues/1214 and supercedes https://github.com/open-component-model/ocm/pull/1217 as a better solution longterm (by getting rid of the old normalization) and shortterm (by achieving full backwards compatibility + introducing a simple test case)

Note that this changes the default normalization algorithm to be `jsonNormalisation/v3` instead of `jsonNormalisation/v1` as it is important for users to migrate as soon as possible.

Skarlso commented on api/ocm/compdesc/norm_test.go:
```diff
@@ -284,4 +293,51 @@ var _ = Describe("Normalization", func() {
+		})
+		It("normalizes v1 with extra identity defaulting", func() {
+			n := Must(compdesc.Normalize(cd, compdesc.JsonNormalisationV1))
+			Expect(string(n)).To(StringEqualTrimmedWithContext("[{\"component\":[{\"componentReferences\":[]},{\"name\":\"ocm.software/duplicate-resource/test\"},{\"provider\":\"internal\"},{\"resources\":[[{\"extraIdentity\":[{\"version\":\"1.0.0\"}]},{\"name\":\"image\"},{\"relation\":\"local\"},{\"type\":\"ociImage\"},{\"version\":\"1.0.0\"}],[{\"extraIdentity\":null},{\"name\":\"image\"},{\"relation\" ...
```
How come the second extraIdentity is `null`? Should that be `2.0.0`?

Skarlso replied:
Looking at the implementation now it's clear why it's null.

Skarlso commented on api/ocm/compdesc/normalizations/legacy/legacy.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+func defaultingOfVersionIntoExtraIdentity(meta []accessors.ElementMeta) {
+	for i := range meta {
+		for j := range meta {
```
What do you think the frequency of call is here?

If we think this is a point where a lot of times for a lot of resources serialization happens, we should improve this.

Improving it shouldn't be too difficult. We could use a seen := map[string]map[string]struct{} where the second map is the identity digest to compare if we have seen that particular identity for that name we already encountered.

That change would make this a single loop loop. :)

But if you think that's a premature optimization, I don't care. :D

jakobmoellerdev replied:
I can do an identity mapping to get it to O(N), no problem. I think its a good practice even though I dont think it matters all too much (were talking about 100x100 entries in the biggest of components.) Good catch :)

jakobmoellerdev replied:
```

func defaultingOfVersionIntoExtraIdentity(meta []accessors.ElementMeta) {
	// nameToIdentity keeps track of the first occurrence of a given (name+extraIdentity)
	// It maps:
	//   resource name -> (string representation of extra identity -> first index seen)
	nameToIdentity := make(map[string][]int, len(meta))
	// conflicts contains a list of the keys that are conflicting so we can easily get all indices
	conflicts := make(map[string]struct{}, len(meta))

	for i, e := range meta {
		// Stable way to stringify Name+ExtraIdentity.
		key := e.GetName() + e.GetExtraIdentity().String()

		// This is the first time we see this (name + extra identity).
		// Store the index to detect future collisions.
		nameToIdentity[key] = append(nameToIdentity[key], i)

		if len(nameToIdentity[key]) > 1 {
			conflicts[key] = struct{}{}
		}
	}

	for conflict := range conflicts {
		indexes := nameToIdentity[conflict]

		// Overwrite the map entries – we keep the last occurrence as-is
		// and only apply the version to duplicates
		for _, idx := range indexes[:len(indexes)-1] {
			e := meta[idx]

			// A collision is found, so "default the version into extra identity" for the current item.
			eid := e.GetExtraIdentity()
			eid.Set(compdesc.SystemIdentityVersion, e.GetVersion())
			e.GetMeta().SetExtraIdentity(eid)

			Logger.Warn(
				fmt.Sprintf(
					"resource identity duplication was normalized for backwards compatibility, "+
						"to avoid this either specify a unique extra identity per item or switch to %s",
					compdesc.JsonNormalisationV3,
				),
				"name", e.GetName(), "index", idx, "conflicting-indexes", indexes, "extra identity", e.GetExtraIdentity(),
			)
		}
	}
}

```

I will leave this code here in case we need it later on if we have performance issues but its not as readable and I dont think its worth the performance gain, thus resolving. We can easily revisit this later if needed

Skarlso reviewed (commented):
One question, otherwise, lovely job Jakob! :)

mandelsoft reviewed (commented):
I'm not sure, why you created a new normalization v3, which is basically exactly the same as v2, before, and changed v2 to modify the component descriptor.

Normalization should never change anything.

The issue stated, that the problem arises after a transport, therefore it cannot be caused by the normalization, but by the transport handling of the extra identity (as you correctly identified in your comment in the issue).

So, it is caused by the defaulting done in the SetResource and SetSource methods of a CV.  If modify mode is on, it defaults the extra identity according to the new rules. 
This has been intended for the creation of a component version, but this should never happen during a transport.

Therefore, we have to fix the transport behavior, ere. This is only possible by introducing a new modification option for those methods, which instruct the adding to keep the original method data and do not the defaulting. The transport tool then has to give this option. (see #1223)

jakobmoellerdev commented:
> I'm not sure, why you created a new normalization v3, which is basically exactly the same as v2, before, and changed v2 to modify the component descriptor.
> 
> Normalization should never change anything.

I changed it because your change broke existing logic for normalization. Before, it didnt matter if the component versions contained extra identities or not, now it does. V3 simply makes this break obvious. 

> 
> The issue stated, that the problem arises after a transport, therefore it cannot be caused by the normalization, but by the transport handling of the extra identity (as you correctly identified in your comment in the issue).
> 
> So, it is caused by the defaulting done in the SetResource and SetSource methods of a CV. If modify mode is on, it defaults the extra identity according to the new rules. This has been intended for the creation of a component version, but this should never happen during a transport.

This is true, but at the same time, the normalization is not allowed to change. I still think the change you submitted is relevant, but the normalization must stay the same.

> Therefore, we have to fix the transport behavior, ere. This is only possible by introducing a new modification option for those methods, which instruct the adding to keep the original method data and do not the defaulting. The transport tool then has to give this option. (see #1223)

Also true, but existing (already transferred) component versions have to be verifiable as well, even with the transport having changed the extra identity
